openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +61,111 @@ def create(cls, name, course_id, group_type=COHORT):
+        if self.course_user_group.group_type != CourseUserGroup.COHORT:  # pylint: disable=E1101
+            raise ValidationError("CohortMembership cannot be used with CourseGroup types other than COHORT")
+        if self.course_user_group.course_id != self.course_id:
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        self.full_clean(validate_unique=False)
+
+        print "Trying to add user to {}".format(self.course_user_group.name)
+
+        trying_to_save = True
+        while trying_to_save:
+            print "entering loop, trying to get membership for user {user}, course {course}".format(
+                user=self.user.id,  # pylint: disable=E1101
+                course=self.course_id
+            )
+
+            saved_membership = None
+            try:
+                #the following 2 "transaction" lines force a fresh read, they can be removed once we're on django 1.8
+                #http://stackoverflow.com/questions/3346124/how-do-i-force-django-to-ignore-any-caches-and-reload-data
+                with transaction.commit_manually():
+                    transaction.commit()
+
+                saved_membership = CohortMembership.objects.get(
+                    user__id=self.user.id,  # pylint: disable=E1101
+                    course_id=self.course_id
+                )
+            except CohortMembership.DoesNotExist:
+                print "No saved_membership!"
+
+            if saved_membership is not None:
+                if saved_membership.course_user_group == self.course_user_group:
+                    raise ValueError("User {user_name} already present in cohort {cohort_name}".format(
+                        user_name=self.user.username,  # pylint: disable=E1101
+                        cohort_name=self.course_user_group.name
+                    ))
+                self.previous_cohort = saved_membership.course_user_group
+                self.previous_cohort_name = saved_membership.course_user_group.name
+                self.previous_cohort_id = saved_membership.course_user_group.id
+
+            time.sleep(10)
+
+            try:
+                with transaction.commit_on_success():
+                    self.course_user_group.users.add(self.user)  # pylint: disable=E1101
+                    self.user.course_groups.add(self.course_user_group)
+                    if saved_membership is not None:
+                        self.previous_cohort.users.remove(self.user)
+                        self.user.course_groups.remove(self.previous_cohort)
+                        updated = CohortMembership.objects.filter(
+                            id=saved_membership.id,
+                            version=saved_membership.version
+                        ).update(
+                            course_user_group=self.course_user_group,
+                            version=saved_membership.version + 1
+                        )
```

Comment:
Right - what you really need here is an UPSERT, which Django ORM doesn't support without resorting to raw SQL. This solution is clever and might work. Do you know the exact SQL that gets generated by this compound ORM statement? Also, do you know if the same SQL is generated in tests (sqlite) and in stage/prod (MySQL)?